## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [138]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [139]:
# View the column names in the data
df.columns




Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [140]:
# Add a total price column to the data
df['total_price'] = df['qty'] * df['unit_price']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  total_price  
0     1096.80     762.71         7.50  105            1    115164.00  
1       24.95      15.09         1.49   21            0       523.95  
2       13.52       7.86         1.68   39            6       527.28  
3       36.42      24.85         1.23   29            3      1056.18  
4      195.10     108.17        46.43   20            1      3902.00

In [141]:
# Use the describe function to gather some basic statistics
# about the data
df.describe()
df.columns
df.dtypes




first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
total_price    float64
dtype: object

In [142]:
# Remove rows with missing values and round percentages to 2 decimal places
df = df.dropna()
df = df.round(2)
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  total_price  
0     1096.80     762.71         7.50  105            1    115164.00  
1       24.95      15.09         1.49   21            0       523.95  
2       13.52       7.86         1.68   39            6       527.28  
3       36.42      24.85         1.23   29            3      1056.18  
4      195.10     108.17        46.43   20            1      3902.00

In [143]:

# Find the top 3 categories with the highest total sales
top_categories = df.groupby(['category'])['total_price'].sum().nlargest(3)
top_categories

# Convert the output to dollars 
top_categories = top_categories.map('${:,.2f}'.format)
top_categories














category
consumables    $1,756,408,731.90
furniture      $1,307,140,008.96
software         $446,735,628.76
Name: total_price, dtype: object

In [144]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df.groupby(['category'])['category'].count().nlargest(1)
top_category
top_subcategory = df.groupby(['subcategory'])['subcategory'].count().nlargest(1)
top_subcategory





subcategory
bathroom supplies    6424
Name: subcategory, dtype: int64

In [145]:
# Which five clients had the most entries in the data?
# Output a df with the client id and the number of entries
top_clients = df.groupby(['client_id'])['client_id'].count().nlargest(5)
top_clients
top_clients = pd.DataFrame(top_clients)
top_clients
top_clients.columns = ['Number of Entries']
top_clients
top_clients.index.name = 'Client ID'







In [146]:
# Store the client ids of those top 5 clients in a list.
# Use that list to filter the original data
top_clients_list = top_clients.index.tolist()
top_clients_list
top_clients_df = df[df['client_id'].isin(top_clients_list)]
top_clients_df.head()





first     last                                           job  \
5    Jessica    Reyes  Production designer, theatre/television/film   
12   Jessica    Reyes  Production designer, theatre/television/film   
137    Bryan    Myers                                Chief of Staff   
214    Bryan    Myers                                Chief of Staff   
216   Angela  Everett                                     Osteopath   

                phone                     email  client_id  order_id  \
5          5016915923  jessica.reyes@grimes.net      33615   1327730   
12         5016915923  jessica.reyes@grimes.net      33615   6059361   
137  444.233.9146x233     bmyers3696@wilson.com      46820   7109502   
214  444.233.9146x233     bmyers3696@wilson.com      46820   8967498   
216  729-620-6192x875    aeverett@rodriguez.com      66037   1203693   

     order_date  order_week  order_year         item_id     category  \
5    2023-05-02          18        2023  BMB23158-88-9A        decor   
12   2023-05-16          20        2023  BDC65078-60-7M  consumables   
137  2023-05-12          19        2023  CXU60946-73-5B  consumables   
214  2023-02-06           6        2023  EMX21429-07-6D    furniture   
216  2023-04-09          14        2023  XUX09951-98-0M  consumables   

           subcategory  unit_price  unit_cost  unit_weight  qty  line_number  \
5               plants      172.84     154.44         2.79   78            0   
12      paper products      229.74     198.38        27.51  568            2   
137  bathroom supplies      136.13     111.47         5.65  118            9   
214             tables        7.98       6.26         7.05  133            3   
216  bathroom supplies      301.87     166.09         0.26  157            2   

     total_price  
5       13481.52  
12     130492.32  
137     16063.34  
214      1061.34  
216     47393.59

In [147]:
# show all data for the first of those five clients
df.loc[df['client_id'] == 'CLIENT_ID_HERE']
df




first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  total_price  
0                1    115164.00  
1                0       523.95  
2                6       527.28  
3                3      1056.18  
4                1      3902.00  
...            ...          ...  
54634            8      2743.29  
54635            0      9709.73  
54636            6     31188.50  
54637            0       165.76  
54638            5      3103.31  

[54639 rows x 19 columns]

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [148]:
# Create a column that calculates the 
# subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  total_price  \
0     1096.80     762.71         7.50  105            1    115164.00   
1       24.95      15.09         1.49   21            0       523.95   
2       13.52       7.86         1.68   39            6       527.28   
3       36.42      24.85         1.23   29            3      1056.18   
4      195.10     108.17        46.43   20            1      3902.00   

    subtotal  
0  115164.00  
1     523.95  
2     527.28  
3    1056.18  
4    3902.00

In [149]:

# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
# Use the weight column to calculate this.
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 if x > 50 else 10)
df.head()
print(df.head())
print()







       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            w

In [150]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price']
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  total_price   subtotal  \
0     762.71         7.50  105            1    115174.00  115164.00   
1      15.09         1.49   21            0       533.95     523.95   
2       7.86         1.68   39            6       537.28     527.28   
3      24.85         1.23   29            3      1066.18    1056.18   
4     108.17        46.43   20            1      3912.00    3902.00   

   shipping_price  
0              10  
1              10  
2              10  
3              10  
4              10  

[5 rows x 21 columns]

In [151]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  total_price   subtotal  shipping_price  \
0         7.50  105            1    115174.00  115164.00              10   
1         1.49   21            0       533.95     523.95              10   
2         1.68   39            6       537.28     527.28              10   
3         1.23   29            3      1066.18    1056.18              10   
4        46.43   20            1      3912.00    3902.00              10   

       cost  
0  80094.55  
1    326.89  
2    316.54  
3    730.65  
4   2173.40  

[5 rows x 22 columns]

In [152]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['cost']
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  total_price   subtotal  shipping_price      cost    profit  
0            1    115174.00  115164.00              10  80094.55  35079.45  
1            0       533.95     523.95              10    326.89    207.06  
2            6       537.28     527.28              10    316.54    220.74  
3            3      1066.18    1056.18              10    730.65    335.53  
4            1      3912.00    3902.00              10   2173.40   1738.60  

[5 rows x 23 columns]

In [153]:
# Update the top_clients_df to include shipping price and profit
top_clients_df = df[df['client_id'].isin(top_clients_list)]
top_clients_df.head()


first     last                                           job  \
5    Jessica    Reyes  Production designer, theatre/television/film   
12   Jessica    Reyes  Production designer, theatre/television/film   
137    Bryan    Myers                                Chief of Staff   
214    Bryan    Myers                                Chief of Staff   
216   Angela  Everett                                     Osteopath   

                phone                     email  client_id  order_id  \
5          5016915923  jessica.reyes@grimes.net      33615   1327730   
12         5016915923  jessica.reyes@grimes.net      33615   6059361   
137  444.233.9146x233     bmyers3696@wilson.com      46820   7109502   
214  444.233.9146x233     bmyers3696@wilson.com      46820   8967498   
216  729-620-6192x875    aeverett@rodriguez.com      66037   1203693   

     order_date  order_week  order_year  ... unit_price unit_cost unit_weight  \
5    2023-05-02          18        2023  ...     172.84    154.44        2.79   
12   2023-05-16          20        2023  ...     229.74    198.38       27.51   
137  2023-05-12          19        2023  ...     136.13    111.47        5.65   
214  2023-02-06           6        2023  ...       7.98      6.26        7.05   
216  2023-04-09          14        2023  ...     301.87    166.09        0.26   

     qty  line_number  total_price   subtotal  shipping_price       cost  \
5     78            0     13491.52   13481.52              10   12056.32   
12   568            2    130502.32  130492.32              10  112689.84   
137  118            9     16073.34   16063.34              10   13163.46   
214  133            3      1071.34    1061.34              10     842.58   
216  157            2     47403.59   47393.59              10   26086.13   

       profit  
5     1435.20  
12   17812.48  
137   2909.88  
214    228.76  
216  21317.46  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [154]:

#Check receipts for accuracy for 3 orders by   

# Order ID 2742071 had a total price of \$152,811.89
# Order ID 2173913 had a total price of \$162,388.71
# Order ID 6128929 had a total price of \$923,441.25
df.loc[df['order_id'] == 2742071]
df.loc[df['order_id'] == 2173913]
df.loc[df['order_id'] == 6128929]
df.loc[df['order_id'] == 2742071, 'total_price'] = 152811.89
df.loc[df['order_id'] == 2173913, 'total_price'] = 162388.71
df.loc[df['order_id'] == 6128929, 'total_price'] = 923441.25

# Print the order ids and total price for the three orders above
df.loc[df['order_id'] == 2742071]
df.loc[df['order_id'] == 2173913]
df.loc[df['order_id'] == 6128929]
print(df.loc[df['order_id'] == 2742071])
print(df.loc[df['order_id'] == 2173913])
print(df.loc[df['order_id'] == 6128929])













          first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   
23840  Kimberly  Blair  Geographical information systems officer   
24671  Kimberly  Blair  Geographical information systems officer   
42470  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
23840  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
24671  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
42470  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  o

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [155]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients_df.groupby(['client_id'])['total_price'].sum()

# Display as dollars with 2 decimal places
top_clients_df.groupby(['client_id'])['total_price'].sum().map('${:,.2f}'.format)







client_id
24741    $70,178,955.25
33615     $5,841,232.11
38378     $8,386,391.15
46820     $7,319,446.03
66037     $7,997,818.38
Name: total_price, dtype: object

In [156]:
# Summarize the data for the top 5 clients and display as dollars
top_clients_df.groupby(['client_id']).agg({'total_price': ['sum', 'mean', 'min', 'max']}).round(2).map('${:,.2f}'.format)





total_price                                     
                      sum         mean     min             max
client_id                                                     
24741      $70,178,955.25  $339,028.77  $10.00  $56,355,185.80
33615       $5,841,232.11   $26,551.06  $11.02   $1,212,512.50
38378       $8,386,391.15   $40,513.97  $10.52   $1,182,707.44
46820       $7,319,446.03   $35,021.27  $10.48   $2,140,941.97
66037       $7,997,818.38   $37,904.35  $16.57   $2,053,327.06

In [157]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
# Percentages should be rounded to 2 decimal places.
# Use the describe function to gather some basic statistics
# about the data
top_clients_df.groupby(['client_id']).agg({'total_price': ['sum', 'mean', 'min', 'max']}).round(2).map('${:,.2f}'.format)



total_price                                     
                      sum         mean     min             max
client_id                                                     
24741      $70,178,955.25  $339,028.77  $10.00  $56,355,185.80
33615       $5,841,232.11   $26,551.06  $11.02   $1,212,512.50
38378       $8,386,391.15   $40,513.97  $10.52   $1,182,707.44
46820       $7,319,446.03   $35,021.27  $10.48   $2,140,941.97
66037       $7,997,818.38   $37,904.35  $16.57   $2,053,327.06

In [158]:
# Sort the updated data by "Total Profit" form highest to lowest
top_clients_df.groupby(['client_id']).agg({'total_price': ['sum', 'mean', 'min', 'max']}).round(2).map('${:,.2f}'.format).sort_values([('total_price', 'sum')], ascending=False)



total_price                                     
                      sum         mean     min             max
client_id                                                     
38378       $8,386,391.15   $40,513.97  $10.52   $1,182,707.44
24741      $70,178,955.25  $339,028.77  $10.00  $56,355,185.80
66037       $7,997,818.38   $37,904.35  $16.57   $2,053,327.06
46820       $7,319,446.03   $35,021.27  $10.48   $2,140,941.97
33615       $5,841,232.11   $26,551.06  $11.02   $1,212,512.50

In [159]:
# Find the profit margin